libraries/plugins/peerplays_sidechain/sidechain_net_handler_bitcoin.cpp

```diff
@@ -615,70 +619,152 @@ void sidechain_net_handler_bitcoin::recreate_primary_wallet() {
+         const auto &prev_sw = std::next(active_sw);
+         if (prev_sw != swi.rend()) {
+            std::string prev_pw_address = prev_sw->addresses.at(sidechain_type::bitcoin);
+            std::string active_pw_address = address;
+            vector<son_info> sons = prev_sw->sons;
+            transfer_all_btc(prev_pw_address, sons, active_pw_address);
+         }
+      }
+   }
+}
 
-            son_wallet_update_operation op;
-            op.payer = GRAPHENE_SON_ACCOUNT;
-            op.son_wallet_id = (*active_sw).id;
-            op.sidechain = sidechain_type::bitcoin;
-            op.address = res.str();
+void sidechain_net_handler_bitcoin::process_deposit(const son_wallet_deposit_object &swdo) {
+   transfer_deposit_to_primary_wallet(swdo);
+}
 
-            proposal_create_operation proposal_op;
-            proposal_op.fee_paying_account = plugin.get_son_object(plugin.get_current_son_id()).son_account;
-            proposal_op.proposed_ops.emplace_back(op_wrapper(op));
-            uint32_t lifetime = (gpo.parameters.block_interval * gpo.active_witnesses.size()) * 3;
-            proposal_op.expiration_time = time_point_sec(database.head_block_time().sec_since_epoch() + lifetime);
+void sidechain_net_handler_bitcoin::process_withdrawal(const son_wallet_withdraw_object &swwo) {
+   transfer_withdrawal_from_primary_wallet(swwo);
+}
+
+static bool has_enough_signatures(const bitcoin_transaction_object &tx_object) {
+   // TODO: Verify with weights calculation
+   bool has_empty = false;
+   for(auto s: tx_object.signatures)
+      has_empty |= s.second.empty();
+   return !has_empty;
+}
 
-            signed_transaction trx = database.create_signed_transaction(plugin.get_private_key(plugin.get_current_son_id()), proposal_op);
+void sidechain_net_handler_bitcoin::process_signing()
+{
+   const auto &idx = plugin.database().get_index_type<proposal_index>().indices().get<by_id>();
+   vector<proposal_id_type> proposals;
+   for (const auto &proposal : idx) {
+      if (proposal.proposed_transaction.operations.size() != 1)
+         continue;
+      if (proposal.proposed_transaction.operations[0].which() != chain::operation::tag<chain::bitcoin_transaction_send_operation>::value) {
+         continue;
+      }
+      bitcoin_transaction_send_operation tx_object = proposal.proposed_transaction.operations[0].get<bitcoin_transaction_send_operation>();
+      // collect signatures
+      auto sons = plugin.get_sons();
+      for (son_id_type son_id : sons) {
+         auto it = tx_object.signatures.find(son_id);
+         if (it == tx_object.signatures.end())
+            continue;
+         if (it->second.empty())
+         {
+            bitcoin_transaction_sign_operation op;
+            son_object s_obj= plugin.get_son_object(son_id);
+            op.payer = s_obj.son_account;
+            op.proposal_id = proposal.id;
+            fc::ecc::private_key k = plugin.get_private_key(son_id);
```

Comment:
If we add signature here, tx_signed for the son rewards is not updated for that son.

Reply:
Yes,  you are right. This code was written before son rewards were merged to the base branch,  and it is possible it is not match the son reward code,  I'll check and fix it.